Add recipe: Emit an evaluation warning only once (#1170)
nix-dot-dev/nix.dev: source/guides/recipes/emit-a-warning-only-once.md

(emit-a-warning-only-once)=
# Emit an evaluation warning only once

The Nix language provides a `warn` function that lets expression authors provide feedback to their code's callers/users.

> Since such warnings are tied to the flow of evaluation, it may happen that the warning you wish to add is trigger too many times.

Reviewer: Simplify and fix typo
```suggestion
Since such warnings are tied to the flow of evaluation the warning may fire more times than intended.
```

> This document shows a technique to solve that problem, and documents its requirements and limitations.

Reviewer: Pure meta-commentary. I would cut it. Replace with a one-sentence statement of the technique itself or drop entirely. The document works without it.

## Working principle

Nix evaluates the root value of an expression file only once.

> We will exploit this property to emit fewer warnings.

Reviewer: No future tense please. We are going to add a technical writing guide. Use present tense to describe how things happen.

    '';
}
```

Since we want widget production to be efficient, we want to phase out `useLegacyMethod`, but without causing unnecessary disruption, so we add a warning:

```nix
{ lib, pkgs }:

{
  makeWidgetScript = args@{ widgetType, useLegacyMethod ? true }:
    lib.warnIf
      (useLegacyMethod &&
        # not explicitly provided by the caller
        !(args ? useLegacyMethod))
      ''
        ACME Inc is in the process of phasing out the legacy widget production method,
        but `makeWidgetScript` is still invoked without disabling the legacy method.
        If this is intentional, and you are not yet ready to upgrade to the efficient method,
        you may specify `useLegacyMethod = true;` to temporarily prolong your use during
        the phase-out.
      ''
      (pkgs.writeScriptBin "make-widget" ''
        #!${pkgs.runtimeShell}
        echo "Just made a" ${lib.escapeShellArg widgetType} ${lib.optionalString (!useLegacyMethod) "efficiently"}
      '');
}
```

This will cause each `makeWidgetScript` invocation to emit a sizable message, every time it is invoked in the old way.

> Now let's use a top level scope to reduce our warnings to just one.

Reviewer: Remove filler intro
```suggestion
Add a top-level ... 
```

We'll start by adding a `let` binding to the top of the file.
The bindings in it will only be evaluated once, as the evaluator will cache the value it leads up to.

```nix
let
  widgetScriptWarning = <...>; # TODO
in
{ lib, pkgs }:

<...> # The rest of the code
```

> Unfortunately, this means we don't have access to the `lib.warn` family of functions, but fortunately we can usually rely on `builtins.warn` nowadays (more on that later).

Reviewer: Don't hedge a fact that has a precise answer.  builtins.warn was introduced in a precise version. This sentence adds vagueness.

As a user i would not know if i can use builtins.warn or not.

Sugggestion: drop the `nowadays (more on that later)` and state facts instead.

Let's see the whole example:

<!-- the comment may seem redundant for instructional purposes, but makes for good copy-pasting -->
```nix
# this let must be top-level; see https://nix.dev/guides/recipes/emit-a-warning-only-once
let
  inherit (builtins) warn seq;
  widgetScriptWarning = warn ''
    ACME Inc is in the process of phasing out the legacy widget production method,
    but `makeWidgetScript` is still invoked without disabling the legacy method.
    If this is intentional, and you are not yet ready to upgrade to the efficient method,
    you may specify `useLegacyMethod = true;` to temporarily prolong your use during
    the phase-out.
  '' null; # Note the `null` here
in
{ lib, pkgs }:

{
  makeWidgetScript = args@{ widgetType, useLegacyMethod ? true }:
    (if useLegacyMethod &&
        # not explicitly provided by the caller
        !(args ? useLegacyMethod)
      then seq widgetScriptWarning
      else x: x
    )
    pkgs.writeScriptBin "make-widget" ''
      #!${pkgs.runtimeShell}
      echo "Just made a" ${lib.escapeShellArg widgetType} ${lib.optionalString (!useLegacyMethod) "efficiently"}
    '';
}
```

> ## Best of both worlds

Reviewer: I would drop this. Most people don't need this.

Use the described technique to provide context in a single warning, and then trigger it right before a brief warning.
This lets you provide context about the call sites without polluting the log as much.

Here is how you may trigger a context warning:

<!-- just the triggering snippet for brevity; we already have plenty of duplicate-ish code -->
```nix
    lib.warnIf
      (useLegacyMethod &&
        # not explicitly provided by the caller
        !(args ? useLegacyMethod))
      (seq widgetScriptWarning "Implicit use of legacy method for widget ${widgetType}; see prior warning.")
      pkgs.writeScriptBin # ...
```

Note that we use `warnIf` again, and the warning message provides a good opportunity for `seq` to trigger our contextual warning.

## Finding call sites

If the evaluator does not produce an unrelated warning first, a quick method to find a call site is to pass `--show-trace --abort-on-warn` to your Nix command invocation.
Otherwise, `--debugger-on-warn` is more suitable, as it lets you resume evaluation after warnings you ignore.

## `builtins.warn` availability

`builtins.warn` was introduced in Nix 2.23.

> Most users have upgraded far beyond 2.22, but if your code is in the upgrade path for users who may not have, use this polyfill:

Reviewer: Just state the polyfill.

```suggestion
builtins.warn was introduced in Nix 2.23. If you support older versions, use this polyfill:
```

Most users have upgraded far beyond 2.22, statement Is not necessary/verified. 
Depends who your users are. Most of open source users? Most nixpkgs users? The sentence works without that claim.

```nix
let
  warn = builtins.warn or builtins.trace;
  # ...
in
# ...
```

> The `. or` operator will take care of the potentially missing `warn`.

Reviewer: will take is future tense. `. or` typo
```suggestion
The `or` operator takes care of the potentially missing `warn`.
```
